presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayToString.java

```diff
@@ -0,0 +1,131 @@
+
+public final class ArrayToString
+        extends ParametricScalar
+{
+    public static final ArrayToString ARRAY_TO_STRING = new ArrayToString();
+    private static final TypeSignature VARCHAR_TYPE_SIGNATURE = parseTypeSignature(StandardTypes.VARCHAR);
+    private static final String FUNCTION_NAME = "array_to_string";
+    private static final Signature SIGNATURE = new Signature(FUNCTION_NAME, ImmutableList.of(comparableTypeParameter("T")), StandardTypes.VARCHAR,
+            ImmutableList.of("array<T>", StandardTypes.VARCHAR, StandardTypes.VARCHAR),
+            false, false);
+
+    @Override
+    public Signature getSignature()
+    {
+        return SIGNATURE;
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Concatenates the elements of the given array using a delimiter and an optional string to replace nulls";
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type type = types.get("T");
+        TypeSignature valueType = type.getTypeSignature();
+        TypeSignature arrayType = parameterizedTypeName(StandardTypes.ARRAY, valueType);
+        MethodHandle methodHandle = methodHandle(ArrayToString.class, "arrayToString", Type.class, ConnectorSession.class, Slice.class, Slice.class, Slice.class);
+        Signature signature = new Signature(FUNCTION_NAME, VARCHAR_TYPE_SIGNATURE, arrayType, VARCHAR_TYPE_SIGNATURE, VARCHAR_TYPE_SIGNATURE);
+
+        return new FunctionInfo(signature, getDescription(), isHidden(), methodHandle.bindTo(type), isDeterministic(), false, ImmutableList.of(false, false, true));
+    }
+
+    public static Slice arrayToString(Type elementType, ConnectorSession session, Slice slice, Slice delimiter, Slice nullReplacement)
+    {
+        if (elementType instanceof ArrayType || elementType instanceof MapType || elementType instanceof RowType) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Unexpected element type %s", elementType.getDisplayName()));
+        }
+
+        Block arrayBlock = readStructuralBlock(slice);
+        int numElements = arrayBlock.getPositionCount();
+        DynamicSliceOutput dos = new DynamicSliceOutput(numElements * 2);
+        Class<?> javaType = elementType.getJavaType();
+
+        for (int i = 0; i < numElements; i++) {
+            if (arrayBlock.isNull(i)) {
+                if (nullReplacement != null) {
+                    dos.appendBytes(nullReplacement);
+                }
+            }
+            else {
+                if (javaType == boolean.class || javaType == long.class || javaType == double.class) {
+                    dos.appendBytes(utf8Slice(elementType.getObjectValue(session, arrayBlock, i).toString()));
```

Comment:
Instead of using `getObjectValue` I think we should lookup the cast from the type to VARCHAR and use that. For example, see how @dain looked it up in his PR: https://github.com/facebook/presto/pull/2752/files#diff-c5f9abfc2482bc9de92d2927c4035f1f
You shouldn't need the bytecode generation he's doing though